File: LeetCode/0806. Number of Lines To Write String/solution.cpp
/**
 * 27 / 27 test cases passed.
 * Runtime: 0 ms
 * Memory Usage: 6.6 MB 
 */
class Solution {
public:
    vector<int> numberOfLines(vector<int>& widths, string s) {
        int k = 0;
        int level = 1;
        for (auto &c : s) {
            int w = widths[c - 'a'];
            if (k + w > 100) {
                k = 0;
                ++ level;
            }
            k += w;
        }
        return {level, k};
    }
};

File: LeetCode/0806. Number of Lines To Write String/solution.rs
/**
 * 27 / 27 test cases passed.
 * Runtime: 0 ms
 * Memory Usage: 2 MB 
 */
impl Solution {
    pub fn number_of_lines(widths: Vec<i32>, s: String) -> Vec<i32> {
        s.bytes().fold(vec![1i32, 0i32], |mut line, width| {
            let w = widths[(width - b'a') as usize];
            line[1] += w;
            if line[1] > 100 {
                line[0] += 1;
                line[1] = w;
            }
            line
        })
    }
}
